my-webxr-app/src/data/DataLayer.tsx

```diff
@@ -68,39 +67,49 @@ export default class DataLayer implements DataAbstractor {
    *
    * If any error occurs during the operation, the method catches the error and returns `false`.
    *
+   * @pre-conditions batchItems is not empty.
    * @param {BatchedDataStream} batchItems - A 2D array of CSV data that is referenced by row.
    * @returns {Promise<boolean>} A promise that resolves to `true` if the operation was successful,
    * and `false` otherwise.
-   *
-   * TODO reset the flag every when user load a different csv file
+   * @post-condition
+   * - The raw data table in the repository contains the new data.
+   * - The column headers of type 'number' will be cast to 'string'
    */
 
   async storeCSV(batchItems: BatchedDataStream): Promise<boolean> {
     try {
-      const transposedData = DataLayer.transposeData(batchItems);
-
-      const promises = transposedData.map(async (column, index) => {
-        let columnName: string;
-        let newValues: RawColumn;
-
-        if (this.isFirstBatch) {
-          columnName = String(column[0]); // Type cast numeric field header to string
-          newValues = column.slice(1);
-          const aColumn = new Column<RawColumn>(columnName, newValues);
-          await this.repository.addColumn(aColumn, TableName.RAW);
-        } else {
-          const columnNames = await this.repository.getCsvColumnNames();
-          columnName = columnNames[index];
-          newValues = column;
+      assert.ok(batchItems.length > 0, 'Batch items must not be empty');
+      // If it is the first batch, create empty columns in the repository
+      if (this.isFirstBatch) {
+        const columnHeaders = batchItems[0].map((header) => header.toString());
+        const promises = columnHeaders.map(async (columnName) => {
+          const newColumn = new Column<NumericColumn>(columnName, []);
+          await this.repository.addColumn(newColumn, TableName.RAW);
+        });
+        await Promise.all(promises);
+
+        // Remove the header row (first row) from the batchItems
+        if (batchItems.length >= 1) {
+          /* eslint-disable no-param-reassign */
+          batchItems = batchItems.slice(1);
+        }
+      }
+
+      if (batchItems.length > 0) {
+        const transposedData = DataLayer.transposeData(batchItems);
+        const columnNames = await this.repository.getCsvColumnNames();
+        const promises = transposedData.map(async (column, index) => {
+          const columnName = columnNames[index];
+          const newValues = column;
+          // push new values to the column in the repository
```

Comment:
would this cause too much of a merge conflict with your PR @Jesster2829 @LongQuanNguyen ? if it does, I'll revert it

Reply:
It may, I am not sure how much it will be. May be wait until Jesse's PR get merged in first